include/mqt-core/dd/Package.hpp

```diff
@@ -194,63 +194,142 @@ class Package {
-    const auto inc = getUniqueTable<Node>().incRef(p);
-    if (inc && p->ref == 1U) {
-      for (const auto& child : p->e) {
-        incRef(child);
-      }
-    }
+    mark.addToRoots(e);
   }
 
   /**
-   * @brief Decrement the reference count of an edge
-   * @details This is the main function for decreasing reference counts within
-   * the DD package. It decreases the reference count of the complex edge weight
-   * as well as the DD node itself. If the node newly becomes dead, meaning
-   * that its reference count hit zero, the reference count of all children is
-   * recursively decreased.
+   * @brief Decrease the DD's reference count and remove it from the tracking
+   * hashset if the count hits zero.
    * @tparam Node The node type of the edge.
-   * @param e The edge to decrease the reference count of
+   * @param e The edge to decrease the reference count of.
+   * @throws std::invalid_argument If the edge is not part of the tracking
+   * hashset.
    */
-  template <class Node> void decRef(const Edge<Node>& e) noexcept {
-    cn.decRef(e.w);
-    const auto& p = e.p;
-    const auto dec = getUniqueTable<Node>().decRef(p);
-    if (dec && p->ref == 0U) {
-      for (const auto& child : p->e) {
-        decRef(child);
+  template <class Node> void decRef(const Edge<Node>& e) {
+    mark.removeFromRoots(e);
+  }
+
+  template <class Node> [[nodiscard]] auto& getRootSet() noexcept {
+    return mark.getRoots<Node>();
+  }
+
+private:
+  struct MarkAndPerform {
+  public:
+    template <class Edge> using RootSet = std::unordered_map<Edge, std::size_t>;
+
+    /// @brief Add to respective root set. Terminals may be added too.
+    template <class Node> void addToRoots(const Edge<Node>& e) noexcept {
+      getRoots<Node>()[e]++;
+    }
+
+    /// @brief Remove from respective root set. Terminals may be removed too.
+    template <class Node> void removeFromRoots(const Edge<Node>& e) {
+      auto& roots = getRoots<Node>();
+      auto it = roots.find(e);
+      if (it == roots.end()) {
+        throw std::invalid_argument("Edge is not part of the root set.");
+      }
+      if (--it->second == 0U) {
+        roots.erase(it);
       }
     }
-  }
 
+    template <class Result, typename Fn> Result execute(Fn& op) noexcept {
+      mark();
+      Result res = op();
+      unmark();
+      return res;
+    }
+
+    /**
+     * @brief Clear all root sets.
+     */
+    void reset() {
+      vRoots.clear();
+      mRoots.clear();
+      dRoots.clear();
+    }
+
+  private:
+    /// @brief Mark edges contained in @p roots.
+    template <class Edge> static void mark(const RootSet<Edge>& roots) {
+      for (auto& [edge, _] : roots) {
+        auto e = edge;
+        e.mark();
+      }
+    }
+
+    /// @brief Unmark edges contained in @p roots.
+    template <class Edge> static void unmark(const RootSet<Edge>& roots) {
+      for (auto& [edge, _] : roots) {
+        auto e = edge;
+        e.unmark();
+      }
+    }
```

Comment:
Just so that I noted it down: I am still not quite sure if this trick of copying the hashmap key is actually valid.

Reply:
Much easier solution: Make `Edge::mark` and `Edge::unmark` const functions 🫠